In [ ]:
!pip install pyspark  #installing the pyspark library

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [ ]:
#Now I will convert this dataframe into a spark dataframe
#Initializing PySpark
from pyspark import SparkContext, SparkConf
from pyspark.sql import SparkSession
from pyspark.sql import SQLContext
# #Spark Config
conf = SparkConf().setAppName("sample_app")
sc = SparkContext(conf=conf)
spark = SparkSession.builder.appName('Test').getOrCreate()
sqlContext = SQLContext(sc)


In [ ]:
spark_df = spark.read.format("csv").option("header", "true").option("inferschema", "true").load("/content/olympics_new")

In [ ]:
spark_df.show()  #here I am using that cleaned dataset that I have analysed using EDA 

+---+--------------------+----+-------------+----+----------+--------------------+----------+------------+-----------+-----------+--------------------+------------------+
|_c0|                Name| age|      country|year|Date_Given|              sports|gold_medal|silver_medal|brone_medal|total_medal|            Name_new|           age_new|
+---+--------------------+----+-------------+----+----------+--------------------+----------+------------+-----------+-----------+--------------------+------------------+
|  0|      Michael Phelps|23.0|United States|2008|2008-08-24|            Swimming|         8|           0|          0|          8|      Michael Phelps|3.1354942159291497|
|  1|      Michael Phelps|19.0|United States|2004|2004-08-29|            Swimming|         6|           0|          2|          8|      Michael Phelps|2.9444389791664403|
|  2|      Michael Phelps|27.0|United States|2012|2012-08-12|            Swimming|         4|           2|          0|          6|      Michael P

In [ ]:
# enableing the Apache Arrow for converting
# Pandas to pySpark DF(DataFrame)
spark.conf.set("spark.sql.execution.arrow.enabled", "true")
  
# Show the DataFrame
spark_df.show(10)

+---+----------------+----+-------------+----+----------+----------+----------+------------+-----------+-----------+----------------+------------------+
|_c0|            Name| age|      country|year|Date_Given|    sports|gold_medal|silver_medal|brone_medal|total_medal|        Name_new|           age_new|
+---+----------------+----+-------------+----+----------+----------+----------+------------+-----------+-----------+----------------+------------------+
|  0|  Michael Phelps|23.0|United States|2008|2008-08-24|  Swimming|         8|           0|          0|          8|  Michael Phelps|3.1354942159291497|
|  1|  Michael Phelps|19.0|United States|2004|2004-08-29|  Swimming|         6|           0|          2|          8|  Michael Phelps|2.9444389791664403|
|  2|  Michael Phelps|27.0|United States|2012|2012-08-12|  Swimming|         4|           2|          0|          6|  Michael Phelps| 3.295836866004329|
|  3|Natalie Coughlin|25.0|United States|2008|2008-08-24|  Swimming|         1|   

In [ ]:
spark_df.select("sports").distinct().count()  #counting the number of unique sports in the column

49

In [ ]:
# Now I will get the  the Olympic Sports/games in this dataset.
spark_df.select("sports").distinct().show(49)

+--------------------+
|              sports|
+--------------------+
|          Gymnastics|
|              Tennis|
|              Boxing|
|Short-Track Speed...|
|          Ice Hockey|
|              Rowing|
|                Judo|
|            Softball|
|             Sailing|
|            Swimming|
|       Alpine Skiing|
|          Basketball|
|            Handball|
| Rhythmic Gymnastics|
|            Biathlon|
|           Triathlon|
|           Badminton|
|           Athletics|
|      Figure Skating|
|    Beach Volleyball|
|          Equestrian|
|            Shooting|
|              Diving|
|            Skeleton|
|         Ski Jumping|
|     Nordic Combined|
|           Wrestling|
|       Speed Skating|
|        Table Tennis|
|           Taekwondo|
|            Canoeing|
|       Weightlifting|
|           Waterpolo|
|            Baseball|
|   Modern Pentathlon|
|        Snowboarding|
|           Bobsleigh|
|             Cycling|
|Cross Country Skiing|
|          Volleyball|
|          

In [ ]:
#Now I will write a PySpark code to plot the total number of medals in  each Olympic Sport/game
import pyspark.sql.functions as f 

total_medals = spark_df.groupBy("sports").sum("total_medal")
total_medals.orderBy(f.desc("sum(total_medal)")).show(49)

+--------------------+----------------+
|              sports|sum(total_medal)|
+--------------------+----------------+
|            Swimming|             765|
|           Athletics|             753|
|              Rowing|             576|
|            Football|             407|
|              Hockey|             386|
|          Ice Hockey|             384|
|            Handball|             351|
|            Canoeing|             333|
|           Waterpolo|             306|
|             Cycling|             306|
|          Basketball|             287|
|          Gymnastics|             282|
|          Volleyball|             278|
|             Fencing|             256|
|           Wrestling|             245|
|                Judo|             224|
|            Baseball|             216|
|             Sailing|             210|
|            Shooting|             192|
|              Boxing|             188|
|          Equestrian|             182|
|       Weightlifting|             180|


In [ ]:
Pdf = spark_df.toPandas ()
Pdf

_c0              Name   age        country  year  Date_Given  \
0        0    Michael Phelps  23.0  United States  2008  2008-08-24   
1        1    Michael Phelps  19.0  United States  2004  2004-08-29   
2        2    Michael Phelps  27.0  United States  2012  2012-08-12   
3        3  Natalie Coughlin  25.0  United States  2008  2008-08-24   
4        4     Aleksey Nemov  24.0         Russia  2000  2000-10-01   
...    ...               ...   ...            ...   ...         ...   
8608  8569     Ilario Di Buò  42.0          Italy  2008  2008-08-24   
8609  8598  Hiroshi Yamamoto  42.0          Japan  2004  2004-08-29   
8610  8604     Ilario Di Buò  42.0          Italy  2000  2000-10-01   
8611  8608     Butch Johnson  42.0  United States  2000  2000-10-01   
8612  8610   Barbara Mensing  42.0        Germany  2000  2000-10-01   

          sports  gold_medal  silver_medal  brone_medal  total_medal  \
0       Swimming           8             0            0            8   
1       Swimming           6             0            2            8   
2       Swimming           4             2            0            6   
3       Swimming           1             2            3            6   
4     Gymnastics           2             1            3            6   
...          ...         ...           ...          ...          ...   
8608     Archery           0             1            0            1   
8609     Archery           0             1            0            1   
8610     Archery           0             1            0            1   
8611     Archery           0             0            1            1   
8612     Archery           0             0            1            1   

              Name_new   age_new  
0       Michael Phelps  3.135494  
1       Michael Phelps  2.944439  
2       Michael Phelps  3.295837  
3     Natalie Coughlin  3.218876  
4        Aleksey Nemov  3.178054  
...                ...       ...  
8608      Ilario Di Bu  3.931826  
8609  Hiroshi Yamamoto  3.713572  
8610      Ilario Di Bu  3.761200  
8611     Butch Johnson  3.806662  
8612   Barbara Mensing  3.663562  

[8613 rows x 13 columns]

In [ ]:
sports_sort1 = Pdf.groupby(Pdf['sports'], as_index=False, sort= True, group_keys=True)['total_medal'].sum()
   
sports_sort_1 = sports_sort1.sort_values(by = 'total_medal',ascending = True)
sports_sort_1 = sports_sort_1.iloc[:,:]
sports_sort_1


sports  total_medal
33                   Skeleton           18
44                  Triathlon           24
43                 Trampoline           24
26          Modern Pentathlon           24
25                       Luge           36
19           Freestyle Skiing           42
6            Beach Volleyball           48
35               Snowboarding           48
27            Nordic Combined           54
34                Ski Jumping           54
17             Figure Skating           54
8                   Bobsleigh           72
42                     Tennis           78
12                    Curling           82
40               Table Tennis           84
28        Rhythmic Gymnastics           84
0               Alpine Skiing           90
3                   Badminton           96
1                     Archery           96
41                  Taekwondo          112
39      Synchronized Swimming          129
36                   Softball          134
7                    Biathlon          138
32  Short-Track Speed Skating          138
37              Speed Skating          140
14                     Diving          144
11       Cross Country Skiing          174
47              Weightlifting          180
15                 Equestrian          182
9                      Boxing          188
31                   Shooting          192
30                    Sailing          210
4                    Baseball          216
24                       Judo          224
48                  Wrestling          245
16                    Fencing          256
45                 Volleyball          278
20                 Gymnastics          282
5                  Basketball          287
13                    Cycling          306
46                  Waterpolo          306
10                   Canoeing          333
21                   Handball          351
23                 Ice Hockey          384
22                     Hockey          386
18                   Football          407
29                     Rowing          576
2                   Athletics          753
38                   Swimming          765

#In this way I got the sports wise medals so that I can visualize them as per the medals won by each sport.

In [ ]:
import plotly.express as px
fig = px.funnel(sports_sort_1, x='sports', y='total_medal')
fig.show()




*  Given is the plot for the total_medals won in each sport.

*  I plotted the funnel chart to plot the sports wise medals won

*   I found that highest Number of medals have been won in Swimming(765)

*   Least number of medals have been won in skeleton sports(18 only)




In [ ]:
spark_df.printSchema()

root
 |-- _c0: integer (nullable = true)
 |-- Name: string (nullable = true)
 |-- age: double (nullable = true)
 |-- country: string (nullable = true)
 |-- year: integer (nullable = true)
 |-- Date_Given: string (nullable = true)
 |-- sports: string (nullable = true)
 |-- gold_medal: integer (nullable = true)
 |-- silver_medal: integer (nullable = true)
 |-- brone_medal: integer (nullable = true)
 |-- total_medal: integer (nullable = true)
 |-- Name_new: string (nullable = true)
 |-- age_new: double (nullable = true)



In [ ]:
spark_df.select("country").distinct().count()

110

In [ ]:
# Now i will find out the total number of medals won by each country in swimming.
total_medalbycountry_swimming = spark_df.filter(spark_df.sports == 'Swimming').groupBy("country").sum("total_medal")
total_medalbycountry_swimming.orderBy(f.desc("sum(total_medal)")).show(110)

+-------------------+----------------+
|            country|sum(total_medal)|
+-------------------+----------------+
|      United States|             267|
|          Australia|             163|
|        Netherlands|              46|
|              Japan|              43|
|             France|              39|
|              China|              35|
|            Germany|              32|
|             Russia|              20|
|              Italy|              16|
|      Great Britain|              11|
|       South Africa|              11|
|             Sweden|               9|
|            Hungary|               9|
|             Brazil|               8|
|            Ukraine|               7|
|           Zimbabwe|               7|
|            Romania|               6|
|             Canada|               5|
|        South Korea|               4|
|             Poland|               3|
|            Austria|               3|
|            Tunisia|               3|
|              Spain|    

In [ ]:
# Let's Find the total number of medals won by each country in Skeleton
total_medalbycountry_skeleton = spark_df.filter(spark_df.sports == 'Skeleton').groupBy("country").sum("total_medal")
total_medalbycountry_skeleton.orderBy(f.desc("sum(total_medal)")).show(110)

+-------------+----------------+
|      country|sum(total_medal)|
+-------------+----------------+
|       Canada|               4|
|Great Britain|               3|
|  Switzerland|               3|
|United States|               3|
|      Germany|               2|
|       Russia|               1|
|      Austria|               1|
|       Latvia|               1|
+-------------+----------------+



In [ ]:
#USA is coming as a strong olympic contestent so let's find out the number of medals that the US won yearly
usa_medals = spark_df.filter(spark_df.country == 'United States').groupBy("year").sum("total_medal")
usa_medals.orderBy("sum(total_medal)").show()

+----+----------------+
|year|sum(total_medal)|
+----+----------------+
|2006|              52|
|2002|              84|
|2010|              97|
|2000|             243|
|2012|             254|
|2004|             265|
|2008|             317|
+----+----------------+



In [ ]:
#Now I have to Find the total number of medals won by each country.
total_medals_c = spark_df.groupBy("country").sum("total_medal")
total_medals_c.orderBy(f.desc("sum(total_medal)")).show(110)

+--------------------+----------------+
|             country|sum(total_medal)|
+--------------------+----------------+
|       United States|            1312|
|              Russia|             768|
|             Germany|             629|
|           Australia|             609|
|               China|             530|
|              Canada|             370|
|               Italy|             331|
|       Great Britain|             322|
|              France|             318|
|         Netherlands|             318|
|         South Korea|             308|
|               Japan|             282|
|              Brazil|             218|
|               Spain|             205|
|              Norway|             192|
|                Cuba|             188|
|              Sweden|             181|
|             Hungary|             145|
|             Ukraine|             143|
|           Argentina|             139|
|             Romania|             123|
|             Finland|             118|


In [ ]:
# Let's find out who was the oldest athlete in the olympics.
spark_df.select("age","Name").distinct().count()

8588

In [ ]:
spark_df2 = spark.read.format("csv").option("header", "true").option("inferschema", "true").load("/content/olympix_data_organized_with_header (1).csv")

In [ ]:
players = spark_df2.select("age","Name","country").distinct().groupBy("Name","country").max("age")
#By mentioning the coutry in group by and select function we can get the athlete and his country in a single code
players.orderBy(f.desc("max(age)")).show(1)


+----------+-------+--------+
|      Name|country|max(age)|
+----------+-------+--------+
|Ian Millar| Canada|      61|
+----------+-------+--------+
only showing top 1 row



In [ ]:
#The oldest athlete is Ian Millar from Canada